### DISCUSSION TITLE
Proxy 에 대해서

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#83] [프론트 기술 질문 공간]
Proxy 란 무엇인가요? 
http 통신에서 프록시의 의미와 역할에 대해서 알아보고 정리해주세요

**Comment by michoball** [comment]
클라이언트와 서버 사이에서 다양한 목적을 위해 사용되어지는 중재자 라고 할 수 있다. 
대게 proxy를 사용하면  서버 측에서 실제 클라이언트의 ip 주소를 알 수 없다.  그래서 몇몇 사용자들은 proxy 우회를 사용해 본래의 ip 주소로는 접근할 수 없는 정보에 접근하고는 한다. ( 국가 코드상 접근 불가한 웹사이트에 접속 -> 누누티비처럼 무료로 영화보기 ? 등등 )

하지만 현대의 인터넷 통신을 위해서 매우 중요한 역할을 한다. 
1. 보안 강화 - 보안에 위협되는 정보를 가진 곳에서 서버로 접근을 시도하면 proxy에서 이를 미리 알아채고 제재할 수 있다. 
2.  인터넷 트래픽 분산 - 서버에 많은 접근이 발생하면 proxy에서 이를 분산해서 트래픽을 줄여줄 수 있다. 
3. cdn - 서버와 물리적으로 먼 곳에서 서버에 요청이 오면 proxy 서버에서 이 요청에 필요한 정보를 미리 가지고 있다가 클라이언트에 전달해 서버까지 접근해서 걸리는 시간을 단축할 수 있다. 
4. 사용자와 관리자 구분 - 서버에 사용자가 접근할 때와 관리자가 접근할 때 다르게 접근할 수 있도록 할 수 있다. 

프록시 서버 구축의 방법론에는 다양한 방식이 있는데 자세히 알아보지는 않았습니다.

**Comment by WooGie911** [comment]
프록시는 클라이언트와 서버 중간에 위치하면서 클라이언트의 모든 HTTP 요청을 받아 서버에 전달합니다.
즉 간단히 말해 클라이언트와 서버 사이의 중개인이라고 할수 있습니다.
또한 프록시 서버는 주로 보안을 위해 사용되며 요청과 응답을 필터링하기 위해 사용됩니다.

자세히 알아보면 프록시는 대표적으로 보안, 속도, ACL, Log/Audit, 접속 우회를 위해 사용됩니다.

(1) 보안: 프록시 서버가 중간에 경유되게 되면 IP를 숨기는 것이 가능하며, 프록시 서버를 방화벽으로 사용하기도 한다(프록시 방화벽). 또한, 익명의 사용자가 서버에 접근하는 것을 막는다.

(2) 속도(캐시): Proxy 서버 중 일부는 프록시 서버에 요청된 내용을 Cache한다. 이 후, 동일 요청이 들어오면 서버에 따로 접속할 필요가 없이 저장된 Cache 자원을 반환한다. 따라서 전송 시간을 절약할 수 있고, 외부 트래픽을 줄임으로써 네트워크 병목 현상도 방지할 수 있다. 이는 서비스의 속도를 높여준다.

(3) ACL : 사이트 접근에 대한 접근 정책을 정의할 수 있다. (ACL = Proxy Server에 접속할 수 있는 범위를 설정하는 옵션)

(4) Log/Audit : 회사 내 직원의 인터넷 사용을 레포팅할 수 있다. 반대로 인트라넷의 사용을 레포팅할 수도 있다.

(5) 지역 네트워크의 제한 우회 : 간혹 한국에서는 접속이 제한이 되는 사이트가 있다. IP를 검사해 한국에서의 접속임을 감지하기 때문인데 이런 경우 프록시 서버를 사용하면 접속을 다른나라로 우회할 수 있게 된다. 우회에 사용할 서버 주소와 포트를 구한 후 '인터넷옵션>연결>LAN설정>프록시서버'에서 서버 주소와 포트를 설정해주면 설정해준 서버에서 접속한 것처럼 속일 수 있기 때문에 접속 제한을 우회할 수 있다.

**Comment by javascriptbiggosoo** [comment]
프락시는 웹 보안, 애플리케이션 통합, 성능 최적화 역할을 하고 있습니다.

클라이언트의 HTTP request를 넘겨받아 대신 서버에 전달 할 수 있습니다.

웹 트래픽 사이의 중개자 역할을 하고 request와 response를 필터링하는 등(애플리케이션 바이러스 검출, 성인 콘텐츠 차단 등) 주로 보안을 위해 쓰입니다.
